### PR TITLE
Improve reference provider and add reference widgets

### PR DESCRIPTION
**Pull request by julien-nc**
The card reference provider is now discoverable and searchable. It supports all the available search providers (boards+cards and comments).

There is a minor flaw (i mean just a weird aspect) of the reference provider design. In this case the provider used to resolve and render cards is supporting all the search providers so we can have one single provider entry in the link picker. I'm fine with that but that can be confusive.

New reference widgets for boards and comments.

There would be a lot of factorization to be done between the card and the comment widget. Just waiting we decide what should be their respective content before factorizing.

Also worth mentioning: I changed the CommentService a bit to make it possible to get a comment from a cardId and a commentId. This lead to a small refactoring of the service.

image

TODO:
* [ ] clarify which icons are used in the widgets

### CODE REVIEW

File: lib/Search/CardSearchResultEntry.php [hunk 1]
```diff
@@ -26,13 +26,22 @@
...
 use OCA\Deck\Db\Card;
 use OCA\Deck\Db\Stack;
 use OCP\Search\SearchResultEntry;
 
 class CardSearchResultEntry extends SearchResultEntry {
 	public function __construct(Board $board, Stack $stack, Card $card, $urlGenerator) {
-		parent::__construct('', $card->getTitle(), $board->getTitle() . ' » ' . $stack->getTitle(), $urlGenerator->linkToRouteAbsolute('deck.page.index') . '#/board/' . $board->getId() . '/card/' . $card->getId(), 'icon-deck');
+		parent::__construct(
+			$urlGenerator->getAbsoluteURL(
+				$urlGenerator->imagePath(Application::APP_ID, 'card.svg')
```
Comment (juliushaertl) [inline lib/Search/CardSearchResultEntry.php:39]: Wondering if this icon is a better match:

https://pictogrammers.com/library/mdi/icon/card-bulleted-outline/

Comment (julien-nc) [reply lib/Search/CardSearchResultEntry.php:39]: :+1: We had the same idea. It's the one I'm using in the reference widget.

**Comment by juliushaertl** [comment]
Seems the composer ocp packages are not automatically bumped for deck anymore since we renamed the default branch to main. Filed https://github.com/nextcloud/office/issues/25 for this but feel free to just update that composer package manually for this pr.

**Review by marcelklehr — APPROVED** [review]
Code looks good. Didn't test.

**Comment by juliushaertl** [comment]
Integration test failures seem related to search of comments ;)

**Comment by julien-nc** [comment]
> Integration test failures seem related to search of comments ;)

I changed a search provider ID and didn't adjust the tests... All good now.

**Comment by julien-nc** [comment]
Rebased on main

**Review by juliushaertl — COMMENTED** [review]
Looks good :+1: Please merge after updating nextcloud/ocp if tests pass :)

**Comment by julien-nc** [comment]
`__construct has more required parameters than parent method` for `OCA\Deck\Model\BoardSummary` and `OCA\Deck\Model\CardDetails`.

It does not sound like an issue to me :grin:. Should we disable/ignore this psalm check?

**Comment by juliushaertl** [comment]
Yep, let's ignore that for this PR.

**Review by juliushaertl — APPROVED** [review]

**Pull request merged**